zebROS_ws/src/controllers_2023/src/elevator_controller.cpp

```diff
@@ -0,0 +1,443 @@
+//HPP CONTENTS INTO CPP
+#ifndef ELEVATOR_CONTROLLER_2023
+#define ELEVATOR_CONTROLLER_2023
+
+//added _2023 to the two lines above ^
+
+#include <ros/ros.h>
+#include <hardware_interface/joint_command_interface.h>
+#include <realtime_tools/realtime_buffer.h>
+#include <controller_interface/multi_interface_controller.h>
+#include <talon_controllers/talon_controller_interface.h> // "
+#include <pluginlib/class_list_macros.h> //to compile as a controller
+#include "controllers_2023_msgs/ElevatorSrv.h"
+
+#include "ddynamic_reconfigure/ddynamic_reconfigure.h"
+double MAX_HEIGHT_VAL = 1.7;
+namespace elevator_controller_2023
+{
+
+class ElevatorCommand_2023
+{
+    public:
+        ElevatorCommand_2023()
+        {
+            position_ = 0;
+        }
+        ElevatorCommand_2023(double position)
+        {
+            position_ = position;
+        }
+        double GetPosition() const
+        {
+            return position_;
+        }
+
+
+    private:
+        double position_;
+};
+//this is the actual controller, so it stores all of the  update() functions and the actual handle from the joint interface
+//if it was only one type, controller_interface::Controller<TalonCommandInterface> here
+class ElevatorController_2023 : public controller_interface::MultiInterfaceController<hardware_interface::TalonCommandInterface>
+{
+    public:
+        ElevatorController_2023()
+        {
+        }
+
+        //should this be hardware_interface::TalonCommandInterface instead? What's the reason to import RobotHW then get CommandInterface from that instead of just importing TalonCommandIface?
+        //answer to my question: the TalonCommandInterface is passed in if it's not a multiInterfaceController, and just one kind of joint is made!
+        virtual bool init(hardware_interface::RobotHW *hw,
+                          ros::NodeHandle             &root_nh,
+                          ros::NodeHandle             &controller_nh) override;
+        virtual void starting(const ros::Time &time) override;
+        virtual void update(const ros::Time &time, const ros::Duration &period) override;
+        virtual void stopping(const ros::Time &time) override;
+
+        bool cmdService(controllers_2023_msgs::ElevatorSrv::Request &req,
+                        controllers_2023_msgs::ElevatorSrv::Response &res);
+
+
+    private:
+        ros::Time last_time_down_;
+        talon_controllers::TalonControllerInterface elevator_joint_; //interface for the talon joint
+
+        realtime_tools::RealtimeBuffer<ElevatorCommand_2023> position_command_; //this is the buffer for percent output commands to be published
+        ros::ServiceServer elevator_service_; //service for receiving commands
+
+        bool zeroed_;
+        bool last_zeroed_;
+        double last_position_;
+        //double last_setpoint_;
+        hardware_interface::TalonMode last_mode_;
+
```

Comment:
Can probably get rid of a lot of these extra blank lines
And similarly, another run of astyle on the file to format it would be good before we merge it